presto-geospatial-toolkit/src/main/java/com/facebook/presto/geospatial/rtree/NeighborEnumeration.java

```diff
@@ -0,0 +1,115 @@
+package com.facebook.presto.geospatial.rtree;
+
+import com.esri.core.geometry.ogc.OGCGeometry;
+import com.facebook.presto.geospatial.GeometryUtils;
+import com.facebook.presto.geospatial.Rectangle;
+import com.facebook.presto.geospatial.rtree.StrRTree.Neighbor;
+import com.facebook.presto.geospatial.rtree.StrRTree.Node;
+
+import java.util.Comparator;
+import java.util.Enumeration;
+import java.util.Optional;
+import java.util.PriorityQueue;
+import java.util.Spliterator;
+import java.util.stream.Stream;
+
+import static com.google.common.base.Preconditions.checkArgument;
+
+public class NeighborEnumeration
+        implements Enumeration<Neighbor>
+{
+    public static final Comparator<Entry> entryComparator = (first, second) -> Double.compare(first.distance, second.distance);
+    private final PriorityQueue<Entry> heap = new PriorityQueue(1, entryComparator);
+    private final Rectangle queryEnvelope;
+    private final OGCGeometry queryGeometry;
+    private Optional<Neighbor> nextCandidate = Optional.empty();
+
+    public NeighborEnumeration(OGCGeometry geometry, Node root)
+    {
+        queryEnvelope = GeometryUtils.getRectangle(geometry);
+        queryGeometry = geometry;
+        heap.add(new Entry(queryEnvelope.distance2(root.getExtent()), root));
+        advance();
+    }
+
+    public boolean hasMoreElements()
+    {
+        return nextCandidate.isPresent();
+    }
+
+    public Neighbor nextElement()
+    {
+        Neighbor neighbor = nextCandidate.get();
+        nextCandidate = Optional.empty();
+        advance();
+        return neighbor;
+    }
+
+    private void advance()
+    {
+        while (!heap.isEmpty()) {
+            Entry entry = heap.poll();
+            Node node = entry.getNode();
+            if (entry.isFinalDistance()) {
+                checkArgument(node.isLeaf(), "Non-leaf node has final distance.");
+                nextCandidate = Optional.of(new Neighbor(Math.sqrt(entry.getDistance()), node.getGeometry().get()));
+                break;
+            }
+            if (node.isLeaf()) {
+                double distance = queryGeometry.distance(node.getGeometry().get());
+                entry.setDistance(distance * distance);
+                entry.setFinalDistance(true);
+                heap.add(entry);
+            }
+            else {
+                node.getChildren().get().stream().forEach(c -> heap.add(new Entry(
+                        queryEnvelope.distance2(c.getExtent()), c
+                )));
+            }
+        }
+    }
+
+    public Stream<Neighbor> asStream()
```

Comment:
If you use an `Iterable` instead of enumeration, then you could use the Guava utility `Streams.stream(iterable)` and wouldn't need this method.